### ISSUE TITLE
[Bug]: expectContextual drops token information and throws generic "unexpected token" error.

### ISSUE THREAD

**Issue by tolmasky** [babel/babel#14144]
### 💻

- [X] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

I ran into this while fixing https://github.com/babel/babel/issues/14123, and the fix for that will include a fix for this too. If you try the following code:

```javascript
require("@babel/parser").parse("import * from 'pizza'", { sourceType: "module" });
```

See it here: https://runkit.com/tolmasky/expectcontextual-drops-token-information

You'll get `SyntaxError: Unexpected token (1:9)`, despite the code having enough information to say "Unexpected token, expect "as"", and in fact it appears like it was originally designed to do so. However, this information is lost between the conversion from `expectContextual` and `unexpected`.

### Current and expected behavior

Should say something like "Unexpected token, expect "as"",

### Environment

- Babel version: 7.16.8
- https://runkit.com/tolmasky/expectcontextual-drops-token-information

**Comment by The-x-Theorist** [comment]
I think that changing the error message can fix this issue.

**Comment by tolmasky** [comment]
I have a fix for this already, it's part of a larger fix for https://github.com/babel/babel/issues/14123. I'm not sure what you mean though, aside from yes, changing the error message would fix the fact that it's the wrong error message. But you need the token information in order to generate a useful error message, which the fix to the above will be able to do now.

**Comment by The-x-Theorist** [comment]
@tolmasky It will be good if this gets fixed in the larger issue, but how're you going to generate the token information?

**Comment by tolmasky** [comment]
The token information is already there, passed to `expectContextual`:

https://github.com/babel/babel/blob/main/packages/babel-parser/src/parser/statement.js#L2526

It just then gets ignored in `expectContextual` since that uses the `ErrorTemplate` form of `this.unexpected()` (instead of the token form). After my change the the token will be able to be used and thus `token` can be passed to `unexpected`. There are several other cases where this happens. If you do a search for `expectContextual` you'll find that 14 uses that could be generating very helpful error messages but instead just do "Unexpected token".

**Comment by The-x-Theorist** [comment]
@tolmasky this looks solved, If you like I can help you fixing this issue in a separate PR?

**Comment by tolmasky** [comment]
It's solved because a different area of the code allowed me to remove the ErrorTemplate form, I am working currently working on adding a bunch more tests for these various things, and it would significantly slow me down to instead try to surgically remove just this one piece. I am aiming to get the umbrella bug PR'ed today. I filed this bug to make it easier to track the various other issues that happened to coincidentally be solved by fixing this other bug, and I marked it as "I would like to fix" and even mentioned the other fix would fix this as well.
